### PR TITLE
Fixes #16081: Automatically restart relayd in case of crash

### PR DESCRIPTION
**Pull request by amousset**
https://issues.rudder.io/issues/16081

### CODE REVIEW

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

File: relay/sources/relayd/src/lib.rs [hunk 1]
```diff
@@ -82,13 +82,19 @@ use tracing_subscriber::{
...
+/// * `Ok(())` means expected shutdown
+/// * `Error(None)` means unexpected crash or panic
+/// * `Error(Some(e))` means shutdown with given error
+pub fn exit_with(r: Result<(), Option<&Error>>) -> ! {
+    exit(match r {
+        Ok(()) => 0,
+        Err(None) => 1,
+        Err(Some(Error::ServerHaltedUnexpectedly)) => 1,
+        Err(Some(Error::ConfigurationParsing(_))) => 2,
+        Err(Some(_)) => 3,
```
Comment (peckpeck) [inline relay/sources/relayd/src/lib.rs]: What are those other errors and why are they considered to be unfixable by a restart ?

Comment (amousset) [reply relay/sources/relayd/src/lib.rs]: Ill update the PR with more details about errors.

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
@peck updated

**Comment by amousset** [comment]
PR updated with a new commit

File: relay/sources/relayd/src/lib.rs [hunk 2]
```diff
@@ -82,12 +82,36 @@ use tracing_subscriber::{
...
+//
+// * Startup, when all config files are loaded, various structures are initialized,
+// and tokio runtime is started. During startup, all errors make the program stop
+// as all steps are necessary to correctly start the app.
+// * Run, once everything is started. Errors are caught and logged, except
+// for panics which stop the program
+//
+// The program should generally only be automatically restarted when
+// encountering an unexpected error (panic). Start errors require external
+// actions.
```
Comment (peckpeck) [inline relay/sources/relayd/src/lib.rs]: Top

**Comment by amousset** [comment]
OK, squash merging this PR

**Pull request merged**